staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -178,13 +184,58 @@ func (q *Type) Done(item interface{}) {
 	if q.dirty.has(item) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
+	} else if q.processing.len() == 0 {
+		q.cond.Signal()
 	}
 }
 
-// ShutDown will cause q to ignore all new items added to it. As soon as the
-// worker goroutines have drained the existing items in the queue, they will be
-// instructed to exit.
+// ShutDown will cause q to ignore all new items added to it and
+// immediately instruct the worker goroutines to exit.
 func (q *Type) ShutDown() {
+	q.setDrain(false)
+	q.shutdown()
+}
+
+// ShutDownWithDrain will cause q to ignore all new items added to it. As soon
```

Comment:
It's important to clarify here: is "draining" about processing every item currently in the queue (including dirty items that will be re-added when a worker exits), or is it about getting every worker goroutine to exit?

Reply:
OK, "draining" is about processing every item currently in the queue. I have update the doc bloc with (what I hope is) a better wording for this.

Reply:
Great -- in that case, you need to exercise the case where an item is added while it's being processed, I'm pretty sure such items get left in the queue right now.

Reply:
Isn't this what `TestAddWhileProcessing` does? I've update https://github.com/kubernetes/kubernetes/blob/525a2032a4ed5eab44bcbb5f356ebf151c47fa46/staging/src/k8s.io/client-go/util/workqueue/queue_test.go#L113 to use `ShutDownWithDrain` and the test passes. Either that test doesn't test what you want me to test, or there's no problem?

Reply:
Did you verify the queue was empty when ShutDownWithDrain exited?

I'd prefer adding a very short test that exercised the scenario deterministically.

Reply:
> Did you verify the queue was empty when ShutDownWithDrain exited?

Yes

> I'd prefer adding a very short test that exercised the scenario deterministically.

I've modified both `TestAddWhileProcessing` and `TestBasic` to include both shut down mechanisms to the tests and added a verification on the queue length for both tests and shut downs. I did this to `TestBasic` too, for completeness sake. Let me know if you think this is sufficient / valid.